src/Elastic.Apm/Metrics/MetricsCollector.cs

```diff
@@ -0,0 +1,150 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Elastic.Apm.Api;
+using Elastic.Apm.Config;
+using Elastic.Apm.Logging;
+using Elastic.Apm.Metrics.MetricsProvider;
+using Elastic.Apm.Report;
+using Timer = System.Timers.Timer;
+
+namespace Elastic.Apm.Metrics
+{
+	/// <summary>
+	/// Iterates through a list of <see cref="IMetricsProvider" /> and
+	/// sends the values through an <see cref="IPayloadSender" /> instance.
+	/// It collects the metrics on an interval read from the <see cref="IConfigurationReader" /> which is
+	/// passed to the constructor.
+	/// In case reading a value from an <see cref="IMetricsProvider" /> fails it retries <see cref="MaxTryWithoutSuccess" />
+	/// times,
+	/// after that it prints a log and won't retry anymore. This is to avoid endlessly trying to read values without success.
+	/// </summary>
+	internal class MetricsCollector : IDisposable, IMetricsCollector
+	{
+		internal const int MaxTryWithoutSuccess = 5;
+
+		private static int _syncPoint;
+
+		/// <summary>
+		/// List of all providers that can provide metrics values.
+		/// Add new providers to this list in case you want the agent to collect more metrics
+		/// </summary>
+		internal readonly List<IMetricsProvider> MetricsProviders;
+
+		private readonly IApmLogger _logger;
+
+		private readonly IPayloadSender _payloadSender;
+
+		private readonly Timer _timer;
+
+		public MetricsCollector(IApmLogger logger, IPayloadSender payloadSender, IConfigurationReader configurationReader)
+		{
+			_logger = logger.Scoped(nameof(MetricsCollector));
+
+			MetricsProviders = new List<IMetricsProvider>
```

Comment:
Now that some metrics providers might start reading data in their constructors we should not construct them if `interval == 0`.

Reply:
👍 makes sense. Moved it after the `if (interval == 0) { ...// return;}`.